File: apps/mobile/android/app/src/main/java/com/tchat/models/ScreenComponent.kt
/**
 * ScreenComponent.kt
 * TchatApp
 *
 * Created by Claude on 22/09/2024.
 */

package com.tchat.models

import kotlinx.serialization.Serializable
import java.util.*

/**
 * Individual UI screens that correspond to web pages
 * Implements the ScreenComponent entity from data-model.md specification
 */
@Serializable
data class ScreenComponent(
    val id: String = UUID.randomUUID().toString(),
    val name: String,
    val route: String,
    val type: NavigationType,
    val platform: Platform = Platform.ANDROID,
    val webEquivalent: String,
    val requiredData: List<String> = emptyList(),
    val optionalData: List<String> = emptyList(),
    val accessLevel: AccessLevel = AccessLevel.PUBLIC,
    val cacheStrategy: CacheStrategy = CacheStrategy.SESSION,
    val offlineSupport: Boolean = false,

    // State management
    var currentState: ScreenState = ScreenState.LOADING,
    var lastStateChange: Long = System.currentTimeMillis(),
    var errorMessage: String? = null
) {

    // MARK: - Enums

    @Serializable
    enum class NavigationType(val value: String) {
        TAB("tab"),
        MODAL("modal"),
        PUSH("push")
    }

    @Serializable
    enum class Platform(val value: String) {
        IOS("ios"),
        ANDROID("android")
    }

    @Serializable
    enum class AccessLevel(val value: String) {
        PUBLIC("public"),
        AUTHENTICATED("authenticated"),
        PREMIUM("premium")
    }

    @Serializable
    enum class CacheStrategy(val value: String) {
        NONE("none"),
        SESSION("session"),
        PERSISTENT("persistent")
    }

    @Serializable
    enum class ScreenState(val value: String) {
        LOADING("loading"),
        READY("ready"),
        DISPLAYED("displayed"),
        NAVIGATING("navigating"),
        HIDDEN("hidden"),
        ERROR("error"),
        RETRY("retry")
    }

    // MARK: - Validation

    /**
     * Validates the screen component according to specification rules
     */
    @Throws(ScreenComponentException::class)
    fun validate() {
        if (id.isBlank()) {
            throw ScreenComponentException.InvalidId("ID cannot be empty")
        }

        if (name.isBlank()) {
            throw ScreenComponentException.InvalidName("Name cannot be empty")
        }

        if (route.isBlank()) {
            throw ScreenComponentException.InvalidRoute("Route cannot be empty")
        }

        if (!route.startsWith("/")) {
            throw ScreenComponentException.InvalidRoute("Route must start with '/'")
        }

        if (webEquivalent.isBlank()) {
            throw ScreenComponentException.InvalidWebEquivalent("Web equivalent route cannot be empty")
        }

        // Platform-specific route validation
        when (platform) {
            Platform.IOS -> {
                if (route.contains(":")) {
                    throw ScreenComponentException.InvalidRoute("iOS routes should not contain ':' parameters")
                }
            }
            Platform.ANDROID -> {
                // Android allows more flexible routing
            }
        }
    }

    // MARK: - State Transitions

    /**
     * Updates the screen state following valid transitions
     */
    @Throws(ScreenComponentException::class)
    fun updateState(newState: ScreenState, errorMessage: String? = null): ScreenComponent {
        if (!isValidTransition(currentState, newState)) {
            throw ScreenComponentException.InvalidStateTransition(
                "Cannot transition from $currentState to $newState"
            )
        }

        return copy(
            currentState = newState,
            lastStateChange = System.currentTimeMillis(),
            errorMessage = errorMessage
        )
    }

    /**
     * Validates state transitions according to specification
     */
    private fun isValidTransition(from: ScreenState, to: ScreenState): Boolean {
        return when (from to to) {
            ScreenState.LOADING to ScreenState.READY,
            ScreenState.LOADING to ScreenState.ERROR -> true

            ScreenState.READY to ScreenState.DISPLAYED,
            ScreenState.READY to ScreenState.ERROR -> true

            ScreenState.DISPLAYED to ScreenState.NAVIGATING,
            ScreenState.DISPLAYED to ScreenState.HIDDEN,
            ScreenState.DISPLAYED to ScreenState.ERROR -> true

            ScreenState.NAVIGATING to ScreenState.HIDDEN,
            ScreenState.NAVIGATING to ScreenState.ERROR -> true

            ScreenState.HIDDEN to ScreenState.DISPLAYED,
            ScreenState.HIDDEN to ScreenState.LOADING -> true

            ScreenState.ERROR to ScreenState.RETRY -> true
            ScreenState.RETRY to ScreenState.LOADING -> true

            else -> false
        }
    }

    // MARK: - Data Dependencies

    /**
     * Checks if all required data dependencies are available
     */
    fun hasRequiredData(availableData: List<String>): Boolean {
        return requiredData.all { availableData.contains(it) }
    }

    /**
     * Gets the missing required data dependencies
     */
    fun getMissingData(availableData: List<String>): List<String> {
        return requiredData.filter { !availableData.contains(it) }
    }

    /**
     * Checks if screen can be displayed offline
     */
    fun canDisplayOffline(cachedData: List<String>): Boolean {
        return offlineSupport && hasRequiredData(cachedData)
    }

    companion object {

        /**
         * Creates a tab screen component
         */
        fun tab(
            name: String,
            route: String,
            webEquivalent: String,
            accessLevel: AccessLevel = AccessLevel.PUBLIC
        ): ScreenComponent {
            return ScreenComponent(
                name = name,
                route = route,
                type = NavigationType.TAB,
                webEquivalent = webEquivalent,
                accessLevel = accessLevel,
                cacheStrategy = CacheStrategy.PERSISTENT,
                offlineSupport = true
            )
        }

        /**
         * Creates a modal screen component
         */
        fun modal(
            name: String,
            route: String,
            webEquivalent: String,
            requiredData: List<String> = emptyList()
        ): ScreenComponent {
            return ScreenComponent(
                name = name,
                route = route,
                type = NavigationType.MODAL,
                webEquivalent = webEquivalent,
                requiredData = requiredData,
                cacheStrategy = CacheStrategy.SESSION
            )
        }

        /**
         * Creates a push navigation screen component
         */
        fun push(
            name: String,
            route: String,
            webEquivalent: String,
            requiredData: List<String> = emptyList()
        ): ScreenComponent {
            return ScreenComponent(
                name = name,
                route = route,
                type = NavigationType.PUSH,
                webEquivalent = webEquivalent,
                requiredData = requiredData,
                cacheStrategy = CacheStrategy.SESSION
            )
        }

        /**
         * Sample screen components for testing and development
         */
        val samples = listOf(
            tab("Chat", "/chat", "/chat", AccessLevel.AUTHENTICATED),
            tab("Store", "/store", "/store", AccessLevel.PUBLIC),
            tab("Social", "/social", "/social", AccessLevel.AUTHENTICATED),
            tab("Video", "/video", "/video", AccessLevel.PREMIUM),
            tab("More", "/more", "/more", AccessLevel.PUBLIC),
            push("Chat Room", "/chat/room", "/chat/room/:id", listOf("roomId", "userId")),
            modal("Settings", "/settings", "/settings")
        )
    }
}

// MARK: - Exception Types

sealed class ScreenComponentException(message: String) : Exception(message) {
    class InvalidId(message: String) : ScreenComponentException(message)
    class InvalidName(message: String) : ScreenComponentException(message)
    class InvalidRoute(message: String) : ScreenComponentException(message)
    class InvalidWebEquivalent(message: String) : ScreenComponentException(message)
    class InvalidStateTransition(message: String) : ScreenComponentException(message)
    class MissingRequiredData(data: List<String>) : ScreenComponentException(
        "Missing required data: ${data.joinToString(", ")}"
    )
}

File: apps/mobile/ios/Sources/Models/ScreenComponent.swift
//
//  ScreenComponent.swift
//  TchatApp
//
//  Created by Claude on 22/09/2024.
//

import Foundation
import SwiftUI

/// Individual UI screens that correspond to web pages
/// Implements the ScreenComponent entity from data-model.md specification
public struct ScreenComponent: Codable, Identifiable, Equatable, Hashable {

    // MARK: - Types

    public enum NavigationType: String, Codable, CaseIterable {
        case tab = "tab"
        case modal = "modal"
        case push = "push"
    }

    public enum Platform: String, Codable, CaseIterable {
        case ios = "ios"
        case android = "android"
    }

    public enum AccessLevel: String, Codable, CaseIterable {
        case `public` = "public"
        case authenticated = "authenticated"
        case premium = "premium"
    }

    public enum CacheStrategy: String, Codable, CaseIterable {
        case none = "none"
        case session = "session"
        case persistent = "persistent"
    }

    public enum ScreenState: String, Codable, CaseIterable {
        case loading = "loading"
        case ready = "ready"
        case displayed = "displayed"
        case navigating = "navigating"
        case hidden = "hidden"
        case error = "error"
        case retry = "retry"
    }

    // MARK: - Properties

    public let id: String
    public let name: String
    public let route: String
    public let type: NavigationType
    public let platform: Platform
    public let webEquivalent: String
    public let requiredData: [String]
    public let optionalData: [String]
    public let accessLevel: AccessLevel
    public let cacheStrategy: CacheStrategy
    public let offlineSupport: Bool

    // State management
    public var currentState: ScreenState
    public var lastStateChange: Date
    public var errorMessage: String?

    // MARK: - Initialization

    public init(
        id: String = UUID().uuidString,
        name: String,
        route: String,
        type: NavigationType,
        platform: Platform = .ios,
        webEquivalent: String,
        requiredData: [String] = [],
        optionalData: [String] = [],
        accessLevel: AccessLevel = .public,
        cacheStrategy: CacheStrategy = .session,
        offlineSupport: Bool = false,
        currentState: ScreenState = .loading,
        lastStateChange: Date = Date(),
        errorMessage: String? = nil
    ) {
        self.id = id
        self.name = name
        self.route = route
        self.type = type
        self.platform = platform
        self.webEquivalent = webEquivalent
        self.requiredData = requiredData
        self.optionalData = optionalData
        self.accessLevel = accessLevel
        self.cacheStrategy = cacheStrategy
        self.offlineSupport = offlineSupport
        self.currentState = currentState
        self.lastStateChange = lastStateChange
        self.errorMessage = errorMessage
    }

    // MARK: - Validation

    /// Validates the screen component according to specification rules
    public func validate() throws {
        guard !id.isEmpty else {
            throw ScreenComponentError.invalidId("ID cannot be empty")
        }

        guard !name.isEmpty else {
            throw ScreenComponentError.invalidName("Name cannot be empty")
        }

        guard !route.isEmpty else {
            throw ScreenComponentError.invalidRoute("Route cannot be empty")
        }

        guard route.hasPrefix("/") else {
            throw ScreenComponentError.invalidRoute("Route must start with '/'")
        }

        guard !webEquivalent.isEmpty else {
            throw ScreenComponentError.invalidWebEquivalent("Web equivalent route cannot be empty")
        }

        // Platform-specific route validation
        switch platform {
        case .ios:
            guard route.contains(":") == false else {
                throw ScreenComponentError.invalidRoute("iOS routes should not contain ':' parameters")
            }
        case .android:
            break // Android allows more flexible routing
        }
    }

    // MARK: - State Transitions

    /// Updates the screen state following valid transitions
    public mutating func updateState(to newState: ScreenState, errorMessage: String? = nil) throws {
        guard isValidTransition(from: currentState, to: newState) else {
            throw ScreenComponentError.invalidStateTransition("Cannot transition from \(currentState) to \(newState)")
        }

        self.currentState = newState
        self.lastStateChange = Date()
        self.errorMessage = errorMessage
    }

    /// Validates state transitions according to specification
    private func isValidTransition(from: ScreenState, to: ScreenState) -> Bool {
        switch (from, to) {
        case (.loading, .ready), (.loading, .error):
            return true
        case (.ready, .displayed), (.ready, .error):
            return true
        case (.displayed, .navigating), (.displayed, .hidden), (.displayed, .error):
            return true
        case (.navigating, .hidden), (.navigating, .error):
            return true
        case (.hidden, .displayed), (.hidden, .loading):
            return true
        case (.error, .retry):
            return true
        case (.retry, .loading):
            return true
        default:
            return false
        }
    }

    // MARK: - Data Dependencies

    /// Checks if all required data dependencies are available
    public func hasRequiredData(_ availableData: [String]) -> Bool {
        return requiredData.allSatisfy { availableData.contains($0) }
    }

    /// Gets the missing required data dependencies
    public func getMissingData(_ availableData: [String]) -> [String] {
        return requiredData.filter { !availableData.contains($0) }
    }

    /// Checks if screen can be displayed offline
    public func canDisplayOffline(_ cachedData: [String]) -> Bool {
        guard offlineSupport else { return false }
        return hasRequiredData(cachedData)
    }
}

// MARK: - Error Types

public enum ScreenComponentError: LocalizedError {
    case invalidId(String)
    case invalidName(String)
    case invalidRoute(String)
    case invalidWebEquivalent(String)
    case invalidStateTransition(String)
    case missingRequiredData([String])

    public var errorDescription: String? {
        switch self {
        case .invalidId(let message),
             .invalidName(let message),
             .invalidRoute(let message),
             .invalidWebEquivalent(let message),
             .invalidStateTransition(let message):
            return message
        case .missingRequiredData(let data):
            return "Missing required data: \(data.joined(separator: ", "))"
        }
    }
}

// MARK: - Convenience Extensions

extension ScreenComponent {

    /// Creates a tab screen component
    public static func tab(
        name: String,
        route: String,
        webEquivalent: String,
        accessLevel: AccessLevel = .public
    ) -> ScreenComponent {
        return ScreenComponent(
            name: name,
            route: route,
            type: .tab,
            webEquivalent: webEquivalent,
            accessLevel: accessLevel,
            cacheStrategy: .persistent,
            offlineSupport: true
        )
    }

    /// Creates a modal screen component
    public static func modal(
        name: String,
        route: String,
        webEquivalent: String,
        requiredData: [String] = []
    ) -> ScreenComponent {
        return ScreenComponent(
            name: name,
            route: route,
            type: .modal,
            webEquivalent: webEquivalent,
            requiredData: requiredData,
            cacheStrategy: .session
        )
    }

    /// Creates a push navigation screen component
    public static func push(
        name: String,
        route: String,
        webEquivalent: String,
        requiredData: [String] = []
    ) -> ScreenComponent {
        return ScreenComponent(
            name: name,
            route: route,
            type: .push,
            webEquivalent: webEquivalent,
            requiredData: requiredData,
            cacheStrategy: .session
        )
    }
}

// MARK: - Sample Data

extension ScreenComponent {

    /// Sample screen components for testing and development
    public static let samples: [ScreenComponent] = [
        .tab(name: "Chat", route: "/chat", webEquivalent: "/chat", accessLevel: .authenticated),
        .tab(name: "Store", route: "/store", webEquivalent: "/store", accessLevel: .public),
        .tab(name: "Social", route: "/social", webEquivalent: "/social", accessLevel: .authenticated),
        .tab(name: "Video", route: "/video", webEquivalent: "/video", accessLevel: .premium),
        .tab(name: "More", route: "/more", webEquivalent: "/more", accessLevel: .public),
        .push(name: "Chat Room", route: "/chat/room", webEquivalent: "/chat/room/:id", requiredData: ["roomId", "userId"]),
        .modal(name: "Settings", route: "/settings", webEquivalent: "/settings")
    ]
}